x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/rule_versions/zip_rule_versions.ts

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { RuleResponse } from '../../../../../../common/detection_engine/rule_schema';
+import type { PrebuiltRuleAsset } from '../../model/rule_assets/prebuilt_rule_asset';
+import type { RuleVersions } from '../diff/calculate_rule_diff';
+
+export const zipRuleVersions = (
+  installedRules: RuleResponse[],
+  baseRules: PrebuiltRuleAsset[],
+  latestRules: PrebuiltRuleAsset[]
+): Map<string, RuleVersions> => {
+  const baseRulesMap = new Map(baseRules.map((r) => [r.rule_id, r]));
+  const latestRulesMap = new Map(latestRules.map((r) => [r.rule_id, r]));
+  const currentRulesMap = new Map(installedRules.map((r) => [r.rule_id, r]));
+
+  const uniqueRuleIds = new Set([
+    ...Array.from(baseRulesMap.keys()),
+    ...Array.from(latestRulesMap.keys()),
+    ...Array.from(currentRulesMap.keys()),
+  ]);
```

Comment:
I think it's not necessary to create a Set to de-duplicate the `ruleId`s, since the Map will override any repeated value with the last key-pair created anyways.
I guess both ways achieve the same thing, not sure if this could make a difference time-complexity-wise, since the number of rules we'll be dealing is not sufficiently large.

Reply:
The idea here is to gather all available rule IDs to iterate over, as these three structures only partially overlap. There could be missing base rules or rules that are no longer available in the package, meaning no latest version, and so on. Ultimately, we aim to create a structure similar to the following:

```ts
const map = {
  rule_id_1: {
    current: undefined,
    base: undefined,
    target: {},
  },
  rule_id_2: {
    current: {},
    base: {},
    target: undefined,
  },
};
```

But a similar result could probably be achieved by iterating over the three structures one by one.